fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingScrubber.java

```diff
@@ -0,0 +1,359 @@
+import java.util.stream.Collectors;
+
+/**
+ * Manifesto:
+ *  Scrub a readable index to validate it's consistency. Repair when needed.
+ *  Keep the type stamp simple, to allow parallel scrubbing.
+ *  Support two states - detect missing and detect dangling. Set by the policy.
+ *  Use rangeSet.missingRanges only for detecting missing.
+ *  (Add option to start dangling-scrub at point?)
+ *  When done, clear typeStamp and Missing ranges. There is a possibility of clearing while
+ *  another scrubbing process is active, but it should be harmless.
+ */
+
+@API(API.Status.INTERNAL)
+public class IndexingScrubber extends IndexingBase {
+    @Nonnull private static final Logger LOGGER = LoggerFactory.getLogger(IndexingScrubber.class);
+    @Nonnull private static final IndexBuildProto.IndexBuildIndexingStamp myIndexingTypeStamp = compileIndexingTypeStamp();
+    private long scanCounterIndexes = 0;
+    private long scanCounterRecords = 0;
+
+    public IndexingScrubber(@Nonnull final IndexingCommon common,
+                            @Nonnull final OnlineIndexer.IndexingPolicy policy,
+                            @Nonnull final OnlineIndexer.ScrubbingPolicy scrubbingPolicy) {
+        super(common, policy, scrubbingPolicy);
+    }
+
+    @Override
+    List<Object> indexingLogMessageKeyValues() {
+        return Arrays.asList(
+                LogMessageKeys.INDEXING_METHOD, "scrub repair"
+        );
+    }
+
+    @Nonnull
+    @Override
+    IndexBuildProto.IndexBuildIndexingStamp getIndexingTypeStamp(final FDBRecordStore store) {
+        return myIndexingTypeStamp;
+    }
+
+    @Nonnull
+    static IndexBuildProto.IndexBuildIndexingStamp compileIndexingTypeStamp() {
+        return
+                IndexBuildProto.IndexBuildIndexingStamp.newBuilder()
+                        .setMethod(IndexBuildProto.IndexBuildIndexingStamp.Method.SCRUB_REPAIR)
+                        .build();
+    }
+
+    @Nonnull
+    private OnlineIndexer.ScrubbingPolicy spolicy() {
+        assert (scrubbingPolicy != null);// this can never happen, just eliminating compiler warnings
+        return scrubbingPolicy;
+    }
+
+    @Override
+    CompletableFuture<Void> buildIndexInternalAsync() {
+        return getRunner().runAsync(context -> openRecordStore(context)
+                .thenCompose( store ->
+                        context.getReadVersionAsync()
+                                .thenCompose(vignore -> {
+                                    SubspaceProvider subspaceProvider = common.getRecordStoreBuilder().getSubspaceProvider();
+                                    return subspaceProvider.getSubspaceAsync(context)
+                                            .thenCompose(subspace ->
+                                                    scrubIndex(subspaceProvider, subspace, null, null)
+                                                    .thenCompose(ignore -> scrubRecords(subspaceProvider, subspace, null, null))
```

Comment:
It's a little weird that the one scrub scrubs both the index and the records. Is that something we're planning on changing in a future change?

Reply:
Actually it's up to the caller. The `scrubbingPolicy` can be set to scrub Dangling only, Missing only, or both.

Reply:
Oh, I see your point. I suppose a different module for each scrubber type would be cleaner from the indexer's prospective, in the expense of adding (not too many) lines of code at the caller's side. 
Would you like me to make this change in this PR, or is it something that can be changed later?

Reply:
I guess the question is how much we care about backwards compatibility on this new code and how hard it would be to extract that (later). I think I'd err on the side of doing the refactoring _now_ while this is fresh rather than trying to untangle it later, but it does probably mean that it takes a little longer to merge this

Reply:
Ok. I'll push a version with the trivial requested changes fixes, then start working on the refactoring - both this one and the API. If it ends up as a big mess (or a crazy time consumer) we'll be able to reset to the trivial.